hibernate-core/src/main/java/org/hibernate/event/internal/DefaultRefreshEventListener.java

```diff
@@ -173,7 +177,15 @@ public void doAfterTransactionCompletion(boolean success, SharedSessionContractI
 
 		String previousFetchProfile = source.getLoadQueryInfluencers().getInternalFetchProfile();
 		source.getLoadQueryInfluencers().setInternalFetchProfile( "refresh" );
-		Object result = persister.load( id, object, event.getLockOptions(), source );
+
+		// Use the entity's current LockMode if it is greater than event.getLockMode()
+		final LockMode currentLockMode = e == null ? null : e.getLockMode();
+		LockOptions lockOptionsToUse = event.getLockOptions();
+		if ( currentLockMode != null && currentLockMode.greaterThan( event.getLockMode() ) ) {
+			lockOptionsToUse = LockOptions.copy( event.getLockOptions(), new LockOptions() );
+			lockOptionsToUse.setLockMode( currentLockMode );
+		}
+		Object result = persister.load( id, object, lockOptionsToUse, source );
```

Comment:
The statement above will fail if ```lockOptionsToUse.getLockMode() == LockMode.WRITE```, because there is no ```UniqueEntityLoader``` that is appropriate for ```LockMode.WRITE```.

This is not a surprise; Javadoc for ```LockMode.WRITE``` says it is not valid for for loading. [1]

The problem is that this will happen whenever an entity that is inserted or updated gets refreshed,  because the ```LockMode``` for the inserted/updated entity is set to ```LockMode.WRITE```.

Which ```UniqueKeyLoader``` should be used in this case? Without the fix, it is the ```UniqueKeyLoader``` with key "refresh". 

After the entity is refreshed, should the entity's ```LockMode``` still be ```LockMode.WRITE```, or should it have some other value?

[1] https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/LockMode.java#L61-L67

Reply:
If the LockMode is WRITE, we already own the exclusive lock on that row, its impossible for another process to update it meaning that there ought to be no changes to "pull back".  From this perspective we could just skip the refresh.

However, the one use-case we do need to worry about (at least the only one I can think of) are db generated values where the user is not taking advantage of Hibernate's generated-value support.  The JPA-standard way to deal with that is to refresh after the persist/merge.

So I think the general case for currentLockMode == WRITE could be to fallback to READ for the load, but to reset the entry's LockMode to WRITE after.

BTW, wouldn't we have the same problem for PESSIMISTIC_WRITE (and PESSIMISTIC_READ possibly)?

Reply:
There isn't a problem with PESSIMISTIC_WRITE or PESSIMISTIC_READ because ```AbstractEntityPersister#loaders``` has entries for both of those lock modes.

Reply:
I think this can be made much simpler.  Working on an alternate PR for just this last little piece (WRITE)